src/offset.rs

```diff
@@ -490,6 +491,108 @@ where
+{
+    type Item = Vec<<T as IntoIterator>::Item>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let (Some(start), Some(&end)) = (self.offsets.next(), self.offsets.peek()) {
+            let slot_start: usize = start.try_into().expect("offset value should be in range");
+            let slot_end: usize = end.try_into().expect("offset value should be in range");
+
+            assert!(
+                slot_end >= slot_start,
+                "offsets must be monotonically increasing, but a slot's start \
+                offset value {slot_start} is greater than the slot's end offset \
+                value {slot_end}, resulting in a negative slot length"
+            );
+
+            let slot_length = slot_end - slot_start;
+            let taken = self.data.by_ref().take(slot_length).collect::<Vec<_>>();
+
+            assert!(
+                taken.len() == slot_length,
+                "underlying data array does not have enough elements, \
+                expected {} elements for this slot, found {}",
+                slot_length,
+                taken.len()
+            );
+
+            Some(taken)
+        } else {
+            None
+        }
+    }
+}
+
+impl<T, OffsetItem: OffsetElement, Buffer: BufferType> IntoIterator
+    for Offset<T, false, OffsetItem, Buffer>
+where
+    T: IntoIterator,
+    <Buffer as BufferType>::Buffer<OffsetItem>: IntoIterator<Item = OffsetItem>,
+    OffsetIntoIter<T, OffsetItem, Buffer>: Iterator<Item = Vec<<T as IntoIterator>::Item>>,
+{
+    type Item = Vec<<T as IntoIterator>::Item>;
+    type IntoIter = OffsetIntoIter<T, OffsetItem, Buffer>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        OffsetIntoIter {
+            data: self.data.into_iter(),
+            offsets: self.offsets.into_iter().peekable(),
+        }
+    }
+}
+
+impl<T, OffsetItem: OffsetElement, Buffer: BufferType> IntoIterator
+    for Offset<T, true, OffsetItem, Buffer>
+where
+    T: IntoIterator,
+    Bitmap<Buffer>: IntoIterator<Item = bool>,
+    <Buffer as BufferType>::Buffer<OffsetItem>: IntoIterator<Item = OffsetItem>,
+    OffsetIntoIter<T, OffsetItem, Buffer>: Iterator<Item = Vec<<T as IntoIterator>::Item>>,
+{
+    type Item = Option<Vec<<T as IntoIterator>::Item>>;
+    type IntoIter = Map<
+        Zip<
+            <Bitmap<Buffer> as IntoIterator>::IntoIter,
+            <OffsetIntoIter<T, OffsetItem, Buffer> as IntoIterator>::IntoIter,
+        >,
+        fn((bool, Vec<<T as IntoIterator>::Item>)) -> Self::Item,
+    >;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.offsets
+            .validity
+            .into_iter()
+            // We use zip insead of map because a null value may have a
+            // positive slot length. That is, a null value may occupy a
+            // non-empty memory space in the data buffer.
+            // Thus the underlying iterator must be advanced even for null
+            // values as indicated by the validity bitmap.
```

Comment:
Can you add a test for this? I think the easiest way to build an example array is to use `FromIterator` and then flip some validity bits.

Reply:
added one in [fix: use iter map and add test for non-empty nulls](https://github.com/mbrobbel/narrow/pull/186/commits/29a92cf2dbd1e88a57ec30a9db7e56c53cdde02f)